File: lib/models/product.dart
import 'dart:convert';
import 'package:e_commerce_app/models/rating.dart';

class ProductModel {
  final String name;
  final String description;
  final double quantity;
  final double price;
  final List<String> images;
  final String category;
  final String? id;
  final List<Rating>? rating;

  ProductModel(
      {required this.name,
      required this.description,
      required this.quantity,
      required this.price,
      required this.images,
      required this.category,
      this.id,
      this.rating});

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'description': description,
      'quantity': quantity,
      'price': price,
      'images': images,
      'category': category,
      'id': id,
      'rating': rating
    };
  }

  factory ProductModel.fromMap(Map<String, dynamic> map) {
    return ProductModel(
        name: map['name'] ?? '',
        description: map['description'] ?? '',
        quantity: map['quantity']?.toDouble() ?? 0.0,
        price: map['price']?.toDouble() ?? 0.0,
        images: List<String>.from(map['images']),
        category: map['category'] ?? '',
        id: map['_id'],
        rating: map['ratings'] != null
            ? List<Rating>.from(
                map['ratings']?.map(
                  (x) => Rating.fromMap(x),
                ),
              )
            : null);
  }

  String toJson() => json.encode(toMap());

  factory ProductModel.fromJson(String source) =>
      ProductModel.fromMap(jsonDecode(source));
}

File: server/models/product.js
const mongoose = require('mongoose');

// schema = structure of product model
const productSchema = mongoose.Schema({
    name: {
        required: true,
        type: String,
        trim: true
    },
    description: {
        required: true,
        type: String,
        trim: true
    },
    images: [
        {
            type: String,
            require: true,
        }
    ],
    quantity: {
        type: Number,
        require: true
    },
    price: {
        type: Number,
        require: true 
    },
    category: {
        type: String,
        require: true
    },
    // rating
});

const Product = mongoose.model('product', productSchema);
module.exports = { Product, productSchema };